// ts-core/src/Events/EventEmitter.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var BaseObject_1 = require("../BaseObject");
var Event = (function (_super) {
    __extends(Event, _super);
    /**
     * Constructor function.
     *
     * @param topic Topic name of event.
     * @param _params Params that will be passed along event.
     * @param caller Context that trigger the original event.
     */
    function Event(topic, _params, caller) {
        _super.call(this);
        this.topic = topic;
        this._params = _params;
        this.caller = caller;
        this.isStopped = false;
    }
    Object.defineProperty(Event.prototype, "params", {
        /**
         * Magic getter for params.
         *
         * @returns {T}
         */
        get: function () {
            return this._params;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Stop the event from being called.
     *
     * @returns {void}
     */
    Event.prototype.stop = function () {
        this.isStopped = true;
    };
    return Event;
}(BaseObject_1["default"]));
exports.Event = Event;
var EventEmitter = (function (_super) {
    __extends(EventEmitter, _super);
    /**
     * Constructor function
     *
     * @returns {EventEmitter}
     */
    function EventEmitter() {
        _super.call(this);
        this._eventCallbacks = {};
    }
    /**
     * Subscribe to triggered events.
     * @param topics        Which topics to listen, separated by space
     * @param callback      Callback function to execute on trigger.
     * @param context       Context for the callback
     * @param once          Run the callback for emitted event exactly one
     * @returns             {EventEmitter}
     */
    EventEmitter.prototype.on = function (topics, callback, context, once) {
        var _this = this;
        if (once === void 0) { once = false; }
        _.each(topics.split(' '), function (topic) {
            // Get or create event collection
            var callbackArray = _this._eventCallbacks[topic];
            if (!callbackArray) {
                callbackArray = [];
                _this._eventCallbacks[topic] = callbackArray;
            }
            // Push callback
            callbackArray.push({
                topic: topic,
                callback: callback,
                context: context,
                once: once
            });
        });
        return this;
    };
    /**
     * Subscribe to emitted topics exactly once
     * @param topics        Which topics to listen, separated by space
     * @param callback      Callback function to execute on trigger.
     * @param context       Context for the callback
     * @returns             {EventEmitter}
     */
    EventEmitter.prototype.once = function (topics, callback, context) {
        return this.on.apply(this, [topics, callback, context, true]);
    };
    /**
     * Unsubscribe from published topics.
     * @param topics        Which topics to stop listening, seperated by space
     * @param callback      Callback function executed on trigger.
     * @param context       Context for the callback
     * @returns             {EventEmitter}
     */
    EventEmitter.prototype.off = function (topics, callback, context) {
        var _this = this;
        _.each(topics.split(' '), function (topic) {
            var callbackArray = _this._eventCallbacks[topic];
            if (!callbackArray) {
                return;
            }
            if (!callback) {
                delete _this._eventCallbacks[topic];
                return;
            }
            var callbacksToRemove = _.where(callbackArray, context ? {
                callback: callback,
                context: context
            } : { callback: callback });
            callbackArray = _.difference(callbackArray, callbacksToRemove);
            if (callbackArray.length == 0) {
                delete _this._eventCallbacks[topic];
            }
            else {
                _this._eventCallbacks[topic] = callbackArray;
            }
        });
        return this;
    };
    /**
     * Publish event for a topic.
     * TODO: Generic for param bag.
     * ````
     * var emitter = new TSCore.Event.EventEmitter();
     * emitter.trigger('topic', arg1, arg2);
     * ````
     * @param topic         Which topic to trigger.
     * @param args          Arguments to pass along event.
     * @returns             {EventEmitter}
     */
    EventEmitter.prototype.trigger = function (topic, params, caller) {
        var _this = this;
        var callbackArray = this._eventCallbacks[topic];
        if (!callbackArray) {
            return;
        }
        // create event
        var event = new Event(topic, params, caller);
        _.each(callbackArray, function (item) {
            if (event.isStopped) {
                return;
            }
            item.callback.apply(item.context || _this, [event]);
            if (item.once) {
                _this.off(topic, item.callback, item.context);
            }
        });
        return this;
    };
    /**
     * Reset all subscriptions.
     *
     * @returns {EventEmitter}
     */
    EventEmitter.prototype.reset = function () {
        this._eventCallbacks = {};
        return this;
    };
    return EventEmitter;
}(BaseObject_1["default"]));
exports.__esModule = true;
exports["default"] = EventEmitter;
